### PR TITLE
tiling

### PR DESCRIPTION
**Pull request by lsutardja**
@vlad-nn , I have a question about a tiling implementation that I wanted to integrate into degirum_tools. The design is based off of a criteria that Shashi had specified earlier on. He wanted a tile model object that you could interface with as if it were a model, the only difference is specifying the tiling parameters. So to me this means any attribute accessible from a dg.model.Model, should be accessible from the "TileModel". The method I used to do this was to wrap the dg.model.Model object in a TileModel, and I used overrides to __getattr__ and __setattr__ to redirect attributes to the wrapped dg.model.Model. This means small changes in the future to dg.model.Model will not necessarily need changes to TileModel. This "pseudo-type casting" hack, won't pass the github actions type checking. Of course I, could force mypy to ignore by type hinting to Any but that defeats the point of type checking. Regardless of the type checking failing, what are your thoughts to this approach to satisfy Shashi's requirement? Also, in general, I was wondering what you thought of this approach to tiling in the degirum_tools framework, especially because degirum_tools has evolved a lot since I first started writing the tiling code.

You can find an example on the use of the TileModel with two different strategies in `test/simpletile_predict.py.` The way I designed it was to make it easy to change strategies to recombine detections, with SimpleTiling as the most naive strategy (take all detections and perform NMS). WBF variants categorize detections for fusion to improve split detections that occur on overlaps. It definitely recovers detections split on two tiles, and theoretically can recover detections split on four tiles (unit test proves the theoretical, but I have never seen an object split on a corner in the wild). LocalGlobal variants include a full image inference along with the tiling to recover the large object mAP which is reduced in the naive tiling approach.

The overlap selection is a little different than generate_tiles_fixed_*. You only supply one overlap value which corresponds to a minimum overlap. If the aspect ratio between model and the input image is different, you will exceed the minimum overlap percent in one of the dimensions. This is another design aspect Shashi asked for, if the end-user supplies a certain tiling rowxcol try your best to fit it and if you can't then go over the requested overlap percentage. Tiling will always find tiles unless the overlap percentage goes above 100%.

### CODE REVIEW

**Comment by vlad-nn** [comment]
@lsutardja , tiling was already implemented in degirum_tools. You can see example of usage in the following notebook: https://github.com/DeGirum/PySDKExamples/blob/main/examples/specialized/tiled_object_detection.ipynb

If you have any questions, please ask.

File: degirum_tools/tile_compound_models.py [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+                else:
+                    dim = tile_height / model_aspect_ratio
+
+                expand_offsets[0] = dim - tile_width
+            elif tile_aspect_ratio > model_aspect_ratio:
+                # Expand the height
+                if model_aspect_ratio >= 1:
+                    dim = tile_width / model_aspect_ratio
+                else:
+                    dim = tile_width * model_aspect_ratio
```
[the remark this answers is not in the document]
Comment (lsutardja) [reply degirum_tools/tile_compound_models.py]: This should be fixed, in the unit test I now have checks for when:
| Case # | Model Aspect Ratio | Tile Aspect Ratio | Model AR vs Tile AR |
| ------ | ------------------ | ----------------- | ------------------- |
| 1      | < 1                | < 1               | >                   |
| 2      | < 1                | < 1               | <                   |
| 3      | < 1                | > 1               | >                   |
| 4      | > 1                | < 1               | <                   |
| 5      | > 1                | > 1               | >                   |
| 6      | > 1                | > 1               | <                   |

File: degirum_tools/tile_compound_models.py [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+            overlap_percent = float(result1.results[0]['label'].split('_')[1].split('@')[1])
+
+            x, y = result1.results[idx]["bbox"][:2]
+
+            if global_result:
+                for i, r in _reverse_enumerate(result2._inference_results):
+                    if "bbox" not in r:
+                        continue
+                    r["bbox"] = np.add(r["bbox"], [x, y, x, y]).tolist()
+                    if not _is_large_object(r['bbox'], self._large_obj_thr * overlap_percent * 5):
```
[the remark this answers is not in the document]
Comment (lsutardja) [reply degirum_tools/tile_compound_models.py:479]: This is based on heuristics. I found that if you leave a certain sized gap between the large objects and the small objects the mAP improves. If we remove the overlap_precent * 5, the mAP goes up. I think I noticed something visually which made me decide to separate this but I can't remember what it was exactly.

| Metric   | w/o heuristic | w/ heuristic |
| -------- | ------------- | ------------ |
| mAP50:95 | 0.22855956    | 0.2312399    |
| mAP50    | 0.39441915    | 0.39702837   |
| mAP75    | 0.22761936    | 0.23131235   |
| mAP S    | 0.1668112     | 0.16680963   |
| mAP M    | 0.2909403     | 0.29740096   |
| mAP L    | 0.37256801    | 0.38272832   |

I didn't want to add another parameter so I used the overlap percent time a constant to modify the large object threshold.

**Comment by lsutardja** [comment]
tiling_compare_fps.py, tile_compound_eval.py, and tile_compound_predict.py aren't actually tests, just example scripts that should eventually be merged/placed into PySDKExamples. I just included it here temporarily for convenience so you can test the usage.

**Review by khatami-mehrdad — APPROVED** [review]

**Pull request merged**